kairos-prover/kairos-circuit-logic/src/account_trie.rs

```diff
@@ -0,0 +1,376 @@
+
+        let recipient_account = if let Some(recipient_account) = recipient_account.get_mut() {
+            if recipient_account.pubkey != deposit.recipient {
+                return Err(("hash collision detected on recipient account").into());
+            }
+
+            recipient_account
+        } else {
+            recipient_account.or_insert_with(|| Account::new(deposit.recipient.clone(), 0, 0))
+        };
+
+        recipient_account.balance = recipient_account
+            .balance
+            .checked_add(deposit.amount)
+            .ok_or("recipient balance overflow")?;
+
+        Ok(())
+    }
+
+    pub fn withdraw(
+        &mut self,
+        withdrawer: &PublicKey,
+        withdraw: &Withdraw,
+        nonce: u64,
+    ) -> Result<(), TxnErr> {
+        let [withdrawer_hash] = hash_buffers([withdrawer.as_slice()]);
+
+        let mut withdrawer_account = self.txn.entry(&withdrawer_hash)?;
+
+        let withdrawer_account = withdrawer_account
+            .get_mut()
+            .ok_or("withdrawer does not have an account")?;
+
+        if withdrawer_account.pubkey != *withdrawer {
+            return Err(("hash collision detected on withdrawer account").into());
+        }
+
+        withdrawer_account.check_nonce(nonce)?;
+        withdrawer_account.increment_nonce();
+
+        withdrawer_account.balance = withdrawer_account
+            .balance
+            .checked_sub(withdraw.amount)
+            .ok_or("sender balance underflow")?;
+
+        Ok(())
+    }
+}
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Account {
+    pub pubkey: PublicKey,
+    pub balance: u64,
+    // Start at 0. Each transfer or withdrawal's `nonce` must match the account's `nonce`.
+    // Each successful transfer or withdrawal increments the `nonce`.
+    pub nonce: u64,
+}
+
+impl Account {
+    pub fn new(pubkey: PublicKey, balance: u64, nonce: u64) -> Self {
+        Self {
+            pubkey,
+            balance,
+            nonce,
+        }
+    }
+
+    pub fn check_nonce(&self, nonce: u64) -> Result<(), TxnErr> {
+        if self.nonce != nonce {
+            return Err(format!(
+                "nonce mismatch: transaction nonce {nonce} does not match account nonce {}",
+                self.nonce,
+            ));
+        }
+
+        Ok(())
+    }
+
+    pub fn increment_nonce(&mut self) {
+        // TODO this is not a real problem an account will never have 2^64 transactions
+        // To make nonce wrapping safe
+        // we should expire transactions after a certain number of batches.
+        self.nonce = self.nonce.wrapping_add(1);
+    }
+}
+
+impl PortableHash for Account {
+    fn portable_hash<H: PortableUpdate>(&self, hasher: &mut H) {
+        self.pubkey.portable_hash(hasher);
+        self.balance.portable_hash(hasher);
+    }
+}
+
+/// A utility function to hash multiple buffers reusing the same hasher.
```

Comment:
Maybe add a docstring that `out[i] = hash(item[i])`. I had to read the impl to understand it. First I thought that `out[i] = hash(item[1] <> hash(item[i-1]))`